src/models/backup.ts

```diff
@@ -460,3 +459,204 @@ export class Drive implements BackupProvider {
+                    resolve(true);
+                  }
+                } else {
+                  resolve(false);
+                }
+              } catch (error) {
+                console.error(error);
+                reject(error);
+              }
+            }
+            return;
+          };
+          xhr.send();
+        }
+      ))
+    ) {
+      await this.refreshToken();
+    }
+    return localStorage.oneDriveToken;
+  }
+
+  private async refreshToken() {
+    return new Promise(
+      (resolve: (value: boolean) => void, reject: (reason: Error) => void) => {
+        const xhr = new XMLHttpRequest();
+        xhr.open(
+          "POST",
+          "https://login.microsoftonline.com/common/oauth2/v2.0/token"
+        );
+        xhr.setRequestHeader(
+          "Content-Type",
+          "application/x-www-form-urlencoded"
+        );
+        xhr.onreadystatechange = () => {
+          if (xhr.readyState === 4) {
+            if (xhr.status === 401) {
+              localStorage.removeItem("oneDriveRefreshToken");
+              localStorage.oneDriveRevoked = true;
+              resolve(false);
+            }
+            try {
+              const res = JSON.parse(xhr.responseText);
+              if (res.error) {
+                if (res.error === "invalid_grant") {
+                  localStorage.removeItem("oneDriveRefreshToken");
+                  localStorage.oneDriveRevoked = true;
+                }
+                console.error(res.error_description);
+                resolve(false);
+              } else {
+                localStorage.oneDriveToken = res.access_token;
+                resolve(true);
+              }
+            } catch (error) {
+              console.error(error);
+              reject(error);
+            }
+          }
+          return;
+        };
+        xhr.send(
+          `client_id=${getCredentials().onedrive.client_id}&refresh_token=${
+            localStorage.oneDriveRefreshToken
+          }&client_secret=${encodeURIComponent(
+            getCredentials().onedrive.client_secret
+          )}&grant_type=refresh_token&scope=https%3A%2F%2Fgraph.microsoft.com%2FFiles.ReadWrite.AppFolder%20https%3A%2F%2Fgraph.microsoft.com%2FUser.Read%20offline_access`
+        );
+      }
+    );
+  }
+
+  async upload(encryption: Encryption) {
+    if (localStorage.oneDriveEncrypted === undefined) {
+      localStorage.oneDriveEncrypted = "true";
+    }
+    const exportData = await EntryStorage.backupGetExport(
+      encryption,
+      localStorage.oneDriveEncrypted === "true"
+    );
+    const backup = JSON.stringify(exportData, null, 2);
+
+    const token = await this.getToken();
+    if (!token) {
+      return new Promise((resolve: (value: boolean) => void) => {
+        resolve(false);
+      });
+    }
+
+    return new Promise(
+      (resolve: (value: boolean) => void, reject: (reason: Error) => void) => {
+        if (!token) {
+          resolve(false);
```

Comment:
Suggestion: I prefer to always add return before resolve and reject, because resolve and reject pass the result of the async function, and the function should end there. Always add return where a promise resolved or rejected can avoid possible bugs

```
return resolve(false);
```